README.md

```diff
@@ -114,6 +114,70 @@ TF_REATTACH_PROVIDERS='{"octopus.com/com/octopusdeploy":{"Protocol":"grpc","Prot
 ```
 5. In the same terminal session where you exported the environment variable, execute the Terraform commands you want to debug.
 
+## Testing
+
+The tests under `integration_test.go` verify the Terraform modules under the `terraform` directory run as expected. At a high level, the tests work like this:
+
+* Create a new Octopus Deploy instance and MSSQL in Docker using Test Containers.
+* Create a new blank space using the module in `1-singlespace`.
+* Populate the new space using one of the other modules under the `terraform` directory.
+* Inspect the new space with the Octopus Go client to verify the newly created resources exist and have the expected settings.
+
+These tests are split by [go-test-split-action](https://github.com/hashicorp-forge/go-test-split-action) to run in parallel across multiple GitHub Action workers as part of the build.
+
+To run the tests locally, you must have the following environment variables:
+
+* `LICENSE`: base 64 encoded octopus license XML file
+* `ECR_ACCESS_KEY`: aws access key (optional, and only used for ECR feed tests)
+* `ECR_SECRET_KEY`: aws secret key (optional, and only used for ECR feed tests)
+* `GIT_CREDENTIAL`: github token (optional, and only used for CaC tests)
+* `GIT_USERNAME`: github username (optional, and only used for CaC tests)
+
+By default, the tests spin up an Octopus instance based on the locally pulled `octopusdeploy/octopusdeploy:latest` image. You can override the image and tag with the following environment variables:
+
+* `OCTOTESTIMAGEURL`: The image to use to create the Octopus instance. Defaults to `octopusdeploy/octopusdeploy`. Can be set to `octopusdeploy/linuxoctopus` to test the Octopus cloud builds. Note you must be logged into DockerHub with the correct credentials to pull this private image.
+* `OCTOTESTVERSION`: The version of the image to use. Defaults to `latest`.
+
+The tests can be run in parallel to speed up the test run time. Set the `GOMAXPROCS` environment variable to the number of parallel tests you want to run. For example, to run 4 tests in parallel, set `GOMAXPROCS=4`. Note though that each test creates it own Octopus and MSSQL docker images, so you will need to have enough resources to run the tests in parallel.
+
+The tests work by executing the `terraform` executable against the test modules. By default, the version of the terraform provider that is used for these tests is defined in the file `config.tf`. The version defined in this file is almost always going to be a few revisions old.
```

Comment:
> The version defined in this file is almost always going to be a few revisions old.

Does this have an impact on the reliability of these tests in the pipeline? It seems like we wouldn't be testing the latest behaviour and any changes we make given this.

Reply:
If so any thoughts on how we could improve this? I'm on sharpening and have been getting up to speed with Terraform and the provider. We'll be migrating to the plugin framework and I'm expecting to learn fairly heavily on test coverage for this.

Reply:
The tests done as part of the GitHub Actions workflow set up the override to ensure any integration tests are done with the version of the code that is checked in. So there is no reason to keep updating the versions in the Terraform files. The only reason would be if a new test was added that required a later version of the provider, in which case it would make sense to include the correct version, if only so people copying the files don't try to use new features with other provider versions.

Reply:
https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/blob/main/.github/workflows/integration-tests.yaml#L75 is where the override is setup as part of the automated tests.